### PR TITLE
Split BadgeRewarder into multiple services

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

This PR removes `BadgeRewarder` from `app/labor` and splits it into multiple services in `app/services`.

Some general notes:

1. A central piece of core logic was factored out into a `Badges::Award` service that the other services call.
2. A lot of the code tried to save the user after a badge was awarded. I changed this to use `touch` instead, which I think is what we actually want.
3. I made some minor cleanups along the way, e.g. rewriting queries for the `ActiveRecord` DSL.
4. The PR looks bigger than it actually is since I primarily just moved code from methods to separate services.

## Related Tickets & Documents

https://github.com/forem/InternalProjectPlanning/issues/271

## QA Instructions, Screenshots, Recordings

1. Specs should be green
2. You can try manually running the badges rake tasks to assign badges.

## Added tests?

- [X] Yes: moved, updated

## Added to documentation?

- [X] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?

Maybe some scheduler updates, pending confirmation from @maestromac.

### CODE REVIEW

File: app/workers/badge_achievements/badge_award_worker.rb [hunk 1]
```diff
@@ -5,10 +5,14 @@ class BadgeAwardWorker
     sidekiq_options queue: :high_priority, retry: 10
 
     def perform(usernames, badge_slug, message)
```
Comment (citizen428) [inline app/workers/badge_achievements/badge_award_worker.rb:7]: @mac Can you please help me verify how we award badges? Is most of it done through Heroku scheduler?

1. The first part of this `if` will take care of things that don't need extra arguments (yearly club, tags, 4, 8, and 16 week streak).
2. The second part here directly awards badges, what are we using this for?
3. The Rake tasks awards top seven, fab five, and contributor, how are these scheduled?
4. What about` award_contributor_badges_from_github` (now `Badges::AwardContributorFromGithub`)? This doesn't seem to be covered by any of these cases, are we still using it?

Comment (maestromac) [reply app/workers/badge_achievements/badge_award_worker.rb:7]: @citizen428 Sorry I missed this. We aren't using Heroku Scheduler anymore. Instead, we are using `sidekiq-cron` gem (shoutout to @mstruve !). You can see each individual task schedule in `config/schedule.yml`.
2. I'm not sure quite sure either. I believe it for one-off awards. cc @benhalpern
3. & 4. You can see each individual task schedule in `config/schedule.yml`.

Comment (citizen428) [reply app/workers/badge_achievements/badge_award_worker.rb:7]: Thanks @maestromac. I had actually looked for a file like this, no idea how I missed it 🤷

**Comment by citizen428** [comment]
I updated the schedule now thanks @maestromac!

File: app/services/badges/award_yearly_club.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+          generate_message(i),
+        )
+      end
+    end
+
+    private
+
+    def generate_message(years)
+      "Happy #{SiteConfig.community_name} birthday! " \
+        "Can you believe it's been #{years} #{'year'.pluralize(years)} already?!"
```
Comment (rhymes) [inline app/services/badges/award_yearly_club.rb:32]: same idea about freezing the template

File: config/schedule.yml [hunk 3]
```diff
@@ -21,49 +21,49 @@ award_yearly_club_badges:
   class: "BadgeAchievements::BadgeAwardWorker"
```
Comment (rhymes) [inline config/schedule.yml:21]: **potentially blocking**: will this break if there are serialized workers with the old argument after the restart deploy?

**Review by rhymes — COMMENTED** [review]
What a refactoring! Left a potentially blocking note on `schedule.yml` and a bunch of other notes

**Comment by citizen428** [comment]
Thanks for all the feedback @rhymes. I'll have a look tomorrow, so far I mostly focused on the splitting and only did some very minor cleanups along the way.

File: config/schedule.yml (hunk 3 above)
Comment (citizen428) [reply config/schedule.yml:21]: Most of them run only every 12 hours, one only runs weekly. I don't think there should be too many queued up jobs hanging around. The only exception is `award_contributor_badges_from_github` which runs hourly at 20 minutes past the full hour, but even there we could probably miss one in the worst case and just wait for the next run to award the badges.

Comment (rhymes) [reply config/schedule.yml:21]: Sure! We can also re-run the manually if we need to

**Comment by maestromac** [comment]
@citizen428 yeah it looks like a manual rake task in `lib/tasks/badges.rake`

File: app/workers/badge_achievements/badge_award_worker.rb [hunk 4]
```diff
@@ -5,10 +5,14 @@ class BadgeAwardWorker
...
-      if BadgeRewarder.respond_to?(badge_slug)
-        BadgeRewarder.public_send(badge_slug)
+      if (award_class = "Badges::#{badge_slug.classify}".safe_constantize)
+        award_class.call
```
Comment (maestromac) [inline app/workers/badge_achievements/badge_award_worker.rb:9]: Love it :star_struck:

**Review by maestromac — APPROVED** [review]
Love the refactor! LGTM!

File: app/services/badges/award.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+module Badges
+  class Award
+    def self.call(user_relation, slug, message_markdown)
+      badge_id = Badge.find_by(slug: slug)&.id
+      return unless badge_id
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award.rb:5]: ✅

File: app/services/badges/award_beloved_comment.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    def self.call(comment_count = 25)
+      new(comment_count).call
+    end
+
+    def initialize(comment_count)
+      @comment_count = comment_count
+    end
+
+    def call
+      badge_id = Badge.find_by(slug: BADGE_SLUG)&.id
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_beloved_comment.rb:14]: ✅

File: app/services/badges/award_streak.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    def self.call(weeks)
+      new(weeks).call
+    end
+
+    def initialize(weeks)
+      @weeks = weeks
+    end
+
+    def call
+      badge_id = Badge.find_by(slug: "#{weeks}-week-streak")&.id
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_streak.rb:19]: ✅

File: app/services/badges/award_beloved_comment.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      end
+    end
+
+    private
+
+    attr_reader :comment_count
+
+    def generate_message(comment)
+      "You're famous! " \
+        "[This is the comment](#{URL.comment(comment)}) for which you are being recognized. 😄"
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_beloved_comment.rb:33]: ✅

File: app/services/badges/award_streak.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    def published_x_weeks_ago?(user, num)
+      user.articles.published
+        .where("published_at > ? AND published_at < ?", num.weeks.ago, (num - 1).weeks.ago).any?
+    end
+
+    def generate_message
+      return LONGEST_STREAK_MESSAGE if weeks == LONGEST_STREAK_WEEKS
+
+      "Congrats on achieving this streak! Consistent writing is hard. " \
+        "The next streak badge you can get is the #{weeks * 2} Week Badge. 😉"
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_streak.rb:55]: ✅

File: app/services/badges/award_yearly_club.rb (hunk 2 above)
Comment (citizen428) [reply app/services/badges/award_yearly_club.rb:32]: ✅

File: app/services/badges/award_eight_week_streak.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+module Badges
+  class AwardEightWeekStreak
+    def self.call
+      ::Badges::AwardStreak.call(8)
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_eight_week_streak.rb:4]: Yes, indeed, I made this change now.

File: app/services/badges/award_contributor_from_github.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
+module Badges
+  class AwardContributorFromGithub
+    BADGE_SLUG = "dev-contributor".freeze
+
+    REPOSITORIES = [
+      "forem/forem",
+      "forem/forem-browser-extension",
+      "thepracticaldev/DEV-Android",
+      "thepracticaldev/DEV-ios",
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_contributor_from_github.rb:9]: ✅

File: app/services/badges/award_fab_five.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+module Badges
+  class AwardFabFive
+    BADGE_SLUG = "fab-5".freeze
+
+    def self.call(usernames, message_markdown = "Congrats!!!")
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_fab_five.rb:5]: ✅

File: app/services/badges/award_top_seven.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+module Badges
+  class AwardTopSeven
+    BADGE_SLUG = "top-7".freeze
+
+    def self.call(usernames, message_markdown = "Congrats!!!")
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_top_seven.rb:5]: ✅

File: app/services/badges/award_tag.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+      end
+    end
+
+    private
+
+    def generate_message(tag, winning_article)
+      "Congratulations on posting the most beloved [##{tag.name}](#{URL.tag(tag)}) post " \
+        "from the past seven days! " \
+        "Your winning post was [#{winning_article.title}](#{URL.article(winning_article)}). " \
+        "(You can only win once per badge-eligible tag)"
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/services/badges/award_tag.rb]: ✅

**Comment by citizen428** [comment]
@rhymes I addressed all of your feedback now.

@maestromac Just making sure you didn't miss [this discussion](https://github.com/forem/forem/pull/11902#discussion_r547212497), as it's potentially relevant for SRE.

**Review by rhymes — APPROVED** [review]
LGTM!

**Comment by maestromac** [comment]
@citizen428 Yep I saw it. It shouldn't be a concern.

**Pull request merged**